packages/backend/src/peripherals/database/VerifierEventRepository.ts

```diff
@@ -0,0 +1,104 @@
+import { Knex } from 'knex'
+import { VerifierEventRow } from 'knex/types/tables'
+
+import { Logger } from '../../tools/Logger'
+import { Repository } from './types'
+
+export type VerifierEventRecord = (
+  | ImplementationAddedEventRecord
+  | UpgradedEventRecord
+) & {
+  id?: number
+}
+
+export interface ImplementationAddedEventRecord {
+  implementation: string
+  initializer: string
+  blockNumber: number
+  name: 'ImplementationAdded'
+}
+
+export interface UpgradedEventRecord {
+  implementation: string
+  blockNumber: number
+  name: 'Upgraded'
+}
+
+export class VerifierEventRepository
```

Comment:
I would change the API surface of this class.
Basically what we need is the following:
- `add(records)`
- `getAll()`
- `getAllBefore(blockNumber)`
- `deleteAllAfter(blockNumber)`
- `deleteAll`
This makes it hard to justify implementing `Repository`, because it requires having `addOrUpdate` which we aren't going to use.

Reply:
What do you feel about this:

```ts
interface Repository<TRecord> {
  addOrUpdate?(records: TRecord[]): Promise<void>
  add?(records: TRecord[]): Promise<void>
}
```

Translating to English, I'd read it as: _If a repository has a method called `add`, then it must take an array of `TRecord`._